### PR TITLE
Add `hamming_weight_encoder` to `models.encodings`

### PR DESCRIPTION
**Pull request by renatomello**
[Reference](https://arxiv.org/abs/2405.20408)

Checklist:
- [ ] Reviewers confirm new code works as expected.
- [x] Tests are passing.
- [x] Coverage does not decrease.
- [x] Documentation is updated.

### CODE REVIEW

**Review by Edoardo-Pedicillo — APPROVED** [review]
Looks good to me

File: src/qibo/models/encodings.py [hunk 1]
```diff
@@ -323,6 +324,106 @@ def unary_encoder_random_gaussian(
...
+    ):
+        targets = list(last_qubit - np.asarray(targets))
+        controls = list(last_qubit - np.asarray(controls))
+        controls.sort()
+
+        if optimize_controls:
+            comparison = k < np.asarray(indices)
+            controls_to_remove = list(np.argwhere(comparison == True).flatten())
+            for control_index in controls_to_remove[::-1]:
+                controls.pop(control_index)
```
Comment (BrunoLiegiBastonLiegi) [inline src/qibo/models/encodings.py:409]: I am not completely sure about this, but can't you simplify this by just taking
```suggestion
            controls = controls[k > np.asarray(indices)]
```
or something along these lines?

File: src/qibo/models/encodings.py [hunk 2]
```diff
@@ -558,3 +659,166 @@ def _parametrized_two_qubit_gate(gate, q0, q1, params=None):
...
+
+
+def _get_markers(bitstring, last_run: bool = False):
+    nqubits = len(bitstring)
+    markers = [len(bitstring) - 1]
+    for ind, value in zip(range(nqubits - 2, -1, -1), bitstring[::-1][1:]):
+        if value == bitstring[-1]:
+            markers.append(ind)
+        else:
+            break
```
Comment (BrunoLiegiBastonLiegi) [inline src/qibo/models/encodings.py:675]: Also here not 100% sure is equivalent, but if I got it right you are checking whether there is another element (except the last one bitstring[-1]) in bitstring with the same value as bitstring[-1], right?
```suggestion
    ind = bitstring[::-1][1:].index(bitstring[-1])
    markers.append(
        list(range(nqubits - 2, -1, -1))[ind]
    )
```

File: tests/test_models_encodings.py [hunk 3]
```diff
@@ -213,6 +217,57 @@ def test_unary_encoder_random_gaussian(backend, nqubits, seed):
...
+        circuit.queue = [
+            gates.X(nqubits - 1 - qubit) for qubit in range(weight)
+        ] + circuit.queue
+    state = backend.execute_circuit(circuit).state()
+
+    backend.assert_allclose(state, target, atol=1e-7)
+
+    # edge case for coverage
+    _ehrlich_algorithm(np.array([0] * nqubits))
+    _get_next_bistring(np.array([0] * nqubits), [], 2)
```
Comment (BrunoLiegiBastonLiegi) [inline tests/test_models_encodings.py:268]: Do these need to be tested as well?

**Review by BrunoLiegiBastonLiegi — APPROVED** [review]
